src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -324,7 +331,12 @@ public override void AfterTransactionCompletion(bool success, ITransaction tx)
 					Factory.StatisticsImplementor.EndTransaction(success);
 				}
 
-				connectionManager.AfterTransaction();
+				// Let the originating session notify the connection manager
+				if (!_transactionCoordinatorShared)
+				{
+					connectionManager.AfterTransaction();
+				}
```

Comment:
In the transaction scope case, the connection manager was getting notified by each session using it. Better let only the originating session do that. This further justifies the ordering: better notify sharing session then originating session last.